### ISSUE TITLE
How does BAT differ from Turing?

### ISSUE THREAD

**Issue by ParadaCarleton** [bat/BAT.jl#321]
Hi! Found this on Github and I'd like to say, BAT.jl looks really interesting, and I'm always happy to see new tools for Bayesian analysis in Julia! I have a quick question -- how does BAT.jl differ from other PPLs in Julia, like Turing, Soss, or Gen?

**Comment by Moelf** [comment]
I want to point out that you can also provide custom likelihood in Turing.jl:
https://turing.ml/dev/docs/using-turing/advanced#update-the-accumulated-log-probability-in-the-model-definition

**Comment by oschulz** [comment]
> you can also provide custom likelihood in Turing.jl

Oh, right, that's a nice feature!

I guess another important difference is that BAT explicitly distinguishes between prior and likelihood, and can support likelihoods that are not (explicitly) tied to data.

**Comment by Moelf** [comment]
I don't know how hard that would be in Turing but surely if you don't use the DSL macro you have more control. Might be worth asking in Slack?

**Comment by oschulz** [comment]
> if you don't use the DSL macro you have more control

Sure, you can of course use the samplers that power Turing directly - which BAT does, e.g. AdvancedHMC. BAT then adds a nice likelihood/prior-oriented interface on top of it, including its parameter space transformation capabilities (which we're currently extending).

**Comment by ParadaCarleton** [comment]
> I guess another important difference is that BAT explicitly distinguishes between prior and likelihood, and can support likelihoods that are not (explicitly) tied to data. The separation between prior and likelihood becomes important for algorithms like MGVI.jl (still to be integrated into BAT), which will also need separation from forward-model and data (we'll add a special likelihood type for that).

Not sure I understand how that differs from Turing -- in DynamicPPL you can distinguish between the prior and the likelihood by working with `tilde_assume` and `tilde_observe`, or by using contexts. Is there some feature missing that would let you use DynamicPPL?

**Comment by oschulz** [comment]
> DynamicPPL you can distinguish between the prior and the likelihood by working

Sure, I didn't mean that DynamicPPL doesn't have a concept of prior and likelihood, but you don't get them as separate objects, right? Though maybe separate prior and likelihood "models" can be composed?




I actually would like to add support for DynamicPPL to BAT, as they're kinda orthogonal. DynamicPPL is a way to express models, whereas BAT is more concerned with taking models from the user and offering a toolset for samling, integration and posterior plotting/analysis.

BAT also differs in doing parameter transformations differently, it transforms the prior to uniform/normal exactly, unlike Bijectors. We're in the process of releasing this as a separate package, so have both transformation mechanisms generally available. This is why we established InverseFunctions.jl and ChangesOfVariables.jl together with the Bijectors developers, to have common APIs.

I don't see BAT and Turing as competing efforts, really  - BAT uses quite a few functions that evolved within Turing, and I hope Turing can make use of functionality well split out as separate packages in the future as well. BAT is currently still too monolithic, the plan is to split it apart an then release BAT v3.0 - hopefully quite soon.

**Comment by Moelf** [comment]
is my understanding correct that in BAT.jl you **have** to define likelihood manually?

**Comment by oschulz** [comment]
> is my understanding correct that in BAT.jl you have to define likelihood manually?

Yes. However, now that we finally have DensityInterface.jl in place, I would like to establish packages that provide likelihoods for standard use cases (e.g. histogram fits, etc.) than can then be used with BAT, MeasureTheory and hopefully other packages as well (Turing could support DensityInterface too via it's "external likelihood" macro, I think).

We'll also add support to BAT for forward models expressed as Markov kernels. I'm not sure in which package to put this yet, it should ideally be somewhere more lightweight than BAT.jl itself. These well be important to support MGVI integration into BAT, but also for automatized posterior predictive checks. It'll look like this:

```julla
prior = Uniform(0, 1)
forward_model(p) = product_distribution(Bernoulli.(p))
likelihood = MarkovModelLikelihood(forward_model, data)
posterior = PosteriorDensity(likelihood, prior)
```

Currently, users still have to provide the full likelihood:

```
prior = Uniform(0, 1)
forward_model(p) = product_distribution(Bernoulli.(p))
likelihood = let data=data
    p -> logpdf(forward_model(p), data)
end
posterior = PosteriorDensity(likelihood, prior)
```

Using `MarkovModelLikelihood` (not the final name) will provide more semantics and allow for calculation of Fisher information and the like. It's fairly high on the priority list, but I'd like to finish splitting up BAT into more modular packages first.

**Comment by oschulz** [comment]
CC @cscherrer
